server/src/main/java/org/opensearch/gateway/AsyncShardsFetchPerNode.java

```diff
@@ -0,0 +1,403 @@
+import org.opensearch.ExceptionsHelper;
+import org.opensearch.OpenSearchTimeoutException;
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.FailedNodeException;
+import org.opensearch.action.support.nodes.BaseNodeResponse;
+import org.opensearch.action.support.nodes.BaseNodesResponse;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.node.DiscoveryNodes;
+import org.opensearch.common.Nullable;
+import org.opensearch.common.lease.Releasable;
+import org.opensearch.index.shard.ShardId;
+import org.opensearch.transport.ReceiveTimeoutTransportException;
+import org.opensearch.common.util.concurrent.OpenSearchRejectedExecutionException;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+
+
+public abstract class AsyncShardsFetchPerNode<T extends BaseNodeResponse> implements Releasable {
+
+    /**
+     * An action that lists the relevant shard data that needs to be fetched.
+     */
+    public interface Lister<NodesResponse extends BaseNodesResponse<NodeResponse>, NodeResponse extends BaseNodeResponse> {
+        void list(DiscoveryNode[] nodes, Map<ShardId,String> shardsIdMap, ActionListener<NodesResponse> listener);
+    }
+
+    protected final Logger logger;
+    protected final String type;
+    protected Map<ShardId,String> shardsToCustomDataPathMap;
+    private final AsyncShardsFetchPerNode.Lister<BaseNodesResponse<T>, T> action;
+    protected final Map<String, AsyncShardsFetchPerNode.NodeEntry<T>> cache = new HashMap<>();
+
+    private final Set<String> nodesToIgnore = new HashSet<>();
+    private final AtomicLong round = new AtomicLong();
+    private boolean closed;
+
+    @SuppressWarnings("unchecked")
+    protected AsyncShardsFetchPerNode(
+        Logger logger,
+        String type,
+        Map<ShardId, String> shardsToCustomDataPathMap,
+        AsyncShardsFetchPerNode.Lister<? extends BaseNodesResponse<T>, T> action
+    ) {
+        this.logger = logger;
+        this.type = type;
+        this.action = (AsyncShardsFetchPerNode.Lister<BaseNodesResponse<T>, T>) action;
+        this.shardsToCustomDataPathMap = shardsToCustomDataPathMap;
+    }
+
+    @Override
+    public synchronized void close() {
+        this.closed = true;
+    }
+
+    protected abstract void reroute(String reason);
+
+    /**
+     * Clear cache for node, ensuring next fetch will fetch a fresh copy.
+     */
+    synchronized void clearCacheForNode(String nodeId) {
+        cache.remove(nodeId);
+    }
+
+    /**
+     * Fills the shard fetched data with new (data) nodes and a fresh NodeEntry, and removes from
+     * it nodes that are no longer part of the state.
+     */
+    private void fillShardCacheWithDataNodes(Map<String, AsyncShardsFetchPerNode.NodeEntry<T>> shardCache, DiscoveryNodes nodes) {
```

Comment:
Cache is still on node level ? 

I'm assuming that the cache object is for every NodeLevelFetcher, so it should contain data for every shard. In that way, shardId should be the new key ..

How can we're managing the map structure same when the whole class is new of a node level fetching ?